lib/parser/context.rb

```diff
@@ -0,0 +1,71 @@
+module Parser
+  # Context of parsing that is represented by a stack of scopes.
+  class Context
+    attr_reader :stack
+
+    def initialize
+      @stack = []
+      freeze
+    end
+
+    def push(state)
+      @stack << state
+    end
+
+    def pop
+      @stack.pop
+    end
+
+    def reset
+      @stack.clear
+    end
+
+    def in_class?
+      @stack.last == :class
+    end
+
+    def in_sclass?
+      @stack.last == :sclass
+    end
+
+    def in_def?
+      @stack.last == :def
+    end
+
+    def in_defs?
+      @stack.last == :defs
+    end
+
+    def in_block?
+      @stack.last == :block
+    end
+
+    def in_lambda?
+      @stack.last == :lambda
+    end
+
+    def indirectly_in_class?
```

Comment:
Do we really need all those? I'm not opposed but it seems extraneous.

Reply:
I've added these methods to make the code self-documented. I agree, it's better to keep only `in_class?/indirectly_in_def?` and add some comments about possibles states.